### PR TITLE
Remove failure

### PR DESCRIPTION
**Pull request by datanel**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -2,12 +2,9 @@
 
 #![deny(missing_docs)]
 
-/// The error type used by the crate.
-pub type Error = failure::Error;
-
-/// The corresponding result type used by the crate.
-pub type Result<T> = std::result::Result<T, Error>;
```
Comment (woshilapin) [inline src/lib.rs]: I would have left that here. This is a `Result` that can be use for the whole module. Imagine for example that we split `Collection` and `CollectionWithId` in two files.

File: src/error.rs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+/// The specific type of an error.
```
Comment (woshilapin) [inline src/error.rs]: This description looks vague to me. How about something like

```suggestion
/// The type of error that the library `collections` might return.
```

File: src/lib.rs [hunk 3]
```diff
@@ -2,12 +2,9 @@
...
-
 mod collection;
+mod error;
+mod macros;
 
 pub use crate::collection::*;
+pub use crate::error::*;
```
[the remark this answers is not in the document]
Comment (datanel) [reply src/lib.rs]: I don't think so. macros are exported automatically.

File: src/error.rs (hunk 2 above)
Comment (datanel) [reply src/error.rs]: I don't know. It's obvious that this this the error collections returns

File: src/lib.rs (hunk 1 above)
Comment (datanel) [reply src/lib.rs]: I already left this line in collection.rs but private.
I'm not sure  if It it should be public.

File: src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (woshilapin) [reply src/lib.rs]: Yes, they are but if do want to do `use collections::impl_id;`, I guess it's not gonna work.

[the remark this answers is not in the document]
Comment (datanel) [reply src/lib.rs]: I already prepared the PR in transit_model for removing `transit_model_collections` without any problems

File: src/lib.rs (hunk 1 above)
Comment (woshilapin) [reply src/lib.rs]: I didn't notice you change visibility of it too. So there is 2 things here:
- should it be public?
- should it be in `lib.rs` or in `collection.rs`?

For the first question, I can go either way. I've seen a [bunch](https://docs.rs/minidom/0.12.0/minidom/error/type.Result.html) [of](http://mvdnes.github.io/rust-docs/zip-rs/zip/result/type.ZipResult.html) [crates](https://docs.serde.rs/serde_json/type.Result.html) [exposing](https://docs.serde.rs/serde_yaml/type.Result.html) [their](https://docs.rs/walkdir/2.3.1/walkdir/type.Result.html) [custom](https://docs.rs/csv/1.1.1/csv/type.Result.html) `Result`. That said, I'm still not convince of the usefulness of that practice.

For the second, see my previous comment. It is a `Result` that we ought to use everywhere in the library. If we create more files that might use it, then I would keep it in `lib.rs`.

File: src/error.rs (hunk 2 above)
Comment (woshilapin) [reply src/error.rs]: Something simpler if you prefer like 
```suggestion
/// Typed error for `collections`.
```

I guess I'm annoyed by `specific` mostly. What does this means?

File: src/lib.rs (hunk 1 above)
Comment (datanel) [reply src/lib.rs]: my point is:
 why expose it?
If we want to expose Result later we can do it without breaking change. 

I don't understand your point about `lib.rs`, we only have one module. We can move the type later in lib.rs if we need more modules.

File: src/error.rs (hunk 2 above)
Comment (datanel) [reply src/error.rs]: specific means it belongs to collections
it's the same as saying `Typed error for `collections` but writing `collections` for me it's pretty obvious . It's not for other crate.

I don't want spend so much time on it so I will change it

**Pull request merged**